rfcs/0008-leaderless-dns.md

```diff
@@ -0,0 +1,238 @@
+
+When a local health check is failing, this IP will not be published to the zone. If it has already been published it will be removed only if there is more than one available address
+
+When a remote health check is failing a TXT record is added/amended in the zone to note the failure - should the percentage of failing clusters exceed the `MultiClusterFailureThreshold` (configured in the DNS Policy), that unhealthy IP is removed from the zone by any cluster with an unhealthy check on that IP.
+
+If all local and remote health checks for a particular host are marked as unhealthy, all known IPs are published, to avoid an NXDOMAIN response.
+
+## OCM / ACM
+This is very similar to a single cluster, with the caveat that the "single" cluster is the hub cluster for this OCM. The primary difference is that the gateway will be a Kuadrant gatewayClass.
+
+### OCM / ACM Health Checks
+This is also the same as single cluster.
+
+## Migrating from Single cluster to Multiple cluster without OCM
+
+In the single cluster: 
+- Edit all relevant provider secrets and set:
+  - `multiClusterID` to some unique value
+
+In the additional clusters:
+- Copy the DNS Policies and provider secrets from the single cluster to any additional clusters, and update the provider secrets with unique `multiClusterID` values.
+- (optional) Create a gateway and workload which uses the same hosts as the single cluster.
+
+## Migrating from single/multiple clusters to OCM
+
+This is not covered.
+
+## Cleanup
+
+Cleanup of a cluster's zone file entries is triggered by deleting the DNS Policy that caused the DNS Records to be created.
+
+The DNS policy will have a finalizer, when it is deleted the controller will mark the created DNS Records for deletion.
+
+The DNS Records will have a finalizer from the DNS Operator which will clean the records from the zone before removing the finalizer.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## General Logic Overview
+
+The general flow for in the Kuadrant operator follows a single path, where it will always output a DNS Record which specifies everything needed for this workload on this cluster, and is unaware of the concept of distributed DNS, or phrased more simply:
+Build a DNS Record (let's refer to this as a kuadrantRecord) based on the local gateways and DNS Policy.
+
+This kuadrantRecord is reconciled by the DNS Operator, where there are 2 possible routes:
+
+Where a multiClusterID is specified in the relevant ProviderSecret the DNS Operator will act with the DNS Provider's zone as the authority and ensure that the records in the zone relevant to the hosts defined within gateways on it's cluster are accurate. When cleaning up a DNS Record in this scenario the operator will ensure all records owned by the policy for the gateway on the local cluster are removed and then prune unresolvable records related to the hosts defined in the gateway from the DNS Provider zone.
```

Comment:
I would vote for a more explicit option of enabling multicluster behavior. maybe something like `multiclusterenabled: true` for enabling and `multiclusterid: asdf` as a configuration parameter. The idea is to prevent accidental activation of multicluster behavior by setting multiclusterid. Also one advantage may also be that it leaves room for signaling of multicluster together with autodetermination of multiclusterid for example by the operator. WDYT?

Reply:
We have a few areas to think about this:
1) "accidental configuration": 
- accidentally setting single-cluster to multi-cluster is not much of a concern, main downside is a bit more traffic to the DNS Provider (really not much with only one cluster though)
- accidentally setting a multi-cluster cluster to single-cluster is quite a concern, the values of other clusters will be erased until this is corrected
2) with a flag as well as a clusterID we see the following scenarios:
- flag: true, clusterID absent: bad configuration, we'd need to prevent this happening somehow (webhook maybe?)
- flag: false, clusterID present: this is bad configuration, but less concerning, however the clusterID is doing nothing here.
- flag: true, clusterID present: good config for multi-cluster
- flag: false, clusterID absent: good config for single-cluster

I think this makes it clear that a flag is only really adding complexity to the configuration without adding any real value when we consider that the clusterID only needs to be non-empty when multi-cluster is active, there is no use-case for the clusterID in a single-cluster or OCM scenario.

I hope that makes my reasoning behind this decision more clear.

Reply:
going to mark as resolved as we agreed on a flag and generating the id

Reply:
As above, we will not be exposing the clusterID to users now, so resolving this comment.